packages/esm-patient-common-lib/src/workspaces/workspaces.ts

```diff
@@ -114,59 +117,75 @@ function getTitleFromExtension(ext: ExtensionRegistration) {
+      workspace.closeWorkspace();
+      // Calling the launchPatientWorkspace again, since one of the `if` case
+      // might resolve, but we need to check all the cases before launching the form.
+      launchPatientWorkspace(name, additionalProps);
+    };
+
+    if (!promptCheckFcn || promptCheckFcn()) {
+      const currentName = workspace.title ?? workspace.name;
       const prompt: Prompt = {
-        title: translateFrom(
-          '@openmrs/esm-patient-chart-app',
-          'activeFormWarning',
-          'There is an active form open in the workspace',
-        ),
+        title: translateFrom('@openmrs/esm-patient-chart-app', 'activeFormWarning', 'You have unsaved changes'),
         body: translateFrom(
           '@openmrs/esm-patient-chart-app',
           'workspaceModalText',
-          `Launching a new form in the workspace could cause you to lose unsaved work on the ${currentName} form.`,
+          'There are unsaved changes in {formName}. Please save them before opening another form.',
           { formName: currentName },
         ),
         onConfirm: () => {
-          const state = store.getState();
           store.setState({
-            openWorkspaces: [newWorkspace, ...state.openWorkspaces.filter((w) => w.type != newWorkspace.type)],
             prompt: null,
           });
+          proceed();
         },
         confirmText: translateFrom('@openmrs/esm-patient-chart-app', 'openAnyway', 'Open anyway'),
       };
-      store.setState({ ...state, prompt });
+      store.setState({ ...store.getState(), prompt });
     } else {
-      const state = store.getState();
-      store.setState({
-        ...state,
-        openWorkspaces: [newWorkspace, ...state.openWorkspaces.filter((w) => w.type != newWorkspace.type)],
-      });
+      proceed();
     }
   }
+
+  const openWorkspaces = store.getState().openWorkspaces;
+  const workspaceIndexInOpenWorkspaces = openWorkspaces.findIndex((w) => w.name === name);
+  const isWorkspaceAlreadyOpen = workspaceIndexInOpenWorkspaces >= 0;
+  const openedWorkspaceWithSameType = openWorkspaces.find((w) => w.type == newWorkspace.type);
+
+  if (openWorkspaces.length === 0) {
+    updateStoreWithNewWorkspace(newWorkspace);
+  } else if (!openWorkspaces[0].canHide && workspaceIndexInOpenWorkspaces !== 0) {
+    promptBeforeLaunchingWorkspace(openWorkspaces[0]);
+  } else if (isWorkspaceAlreadyOpen) {
+    openWorkspaces[workspaceIndexInOpenWorkspaces].additionalProps = newWorkspace.additionalProps;
+    const restOfWorkspaces = openWorkspaces.filter((w) => w.name != name);
+    updateStoreWithNewWorkspace(openWorkspaces[workspaceIndexInOpenWorkspaces], restOfWorkspaces);
+  } else if (!!openedWorkspaceWithSameType) {
+    const restOfWorkspaces = store.getState().openWorkspaces.filter((w) => w.type != newWorkspace.type);
+    updateStoreWithNewWorkspace(openedWorkspaceWithSameType, restOfWorkspaces);
+    promptBeforeLaunchingWorkspace(openedWorkspaceWithSameType);
```

Comment:
I'm not sure I follow what this is supposed to be doing? More specifically, why to we update the store _before_ prompting?

Reply:
As I shared in the pt.4 in the description, if I open the workspace A with type 'clinical-form', then if there is another already opened workspace B with type "clinical-form" and if the B is not visible to the user, I am actually bringing the workspace B to the front, so that the modal to close the workspace B, should have workspace B in front to help the user to exactly know which specific workspace will be closed.

Reply:
Ah! Got it!